fix(vehicle_cmd_gate): fix the problems while transition from/to stopped state

brkay54 opened the pull request:
## Description

<!-- Write a brief description of this PR. -->
This PR solves the following issues:
* If the user sends stop command to autoware while driving, autoware does not do steering control while stopping: https://github.com/autowarefoundation/autoware.universe/issues/5796
* https://github.com/autowarefoundation/autoware.universe/issues/4915
* https://github.com/autowarefoundation/autoware.universe/issues/6044
## Related links

<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->

## Tests performed
Tested on Psim, working as we want.
<!-- Describe how you have tested this PR. -->

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Interface changes

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

## Effects on system behavior

<!-- Describe how this PR affects the system behavior. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

brkay54 commented:
Hi @TakaHoribe, about the problem mentioned this [issue](https://github.com/autowarefoundation/autoware.universe/issues/4915), I moved the steering converge checking from controllers to vehicle_cmd_gate to solve this issue. If this method is okay, please let me know, I will create another PR to remove steering converge checking from controllers.

brkay54 commented:
Hi @TakaHoribe @tkimura4, sorry to disturb you, but just I want to kindly ping you.

tkimura4 commented:
@kosuke55 

Could you review this PR?

> Hi @TakaHoribe, about the problem mentioned this https://github.com/autowarefoundation/autoware.universe/issues/4915, I moved the steering converge checking from controllers to vehicle_cmd_gate to solve this issue. If this method is okay, please let me know, I will create another PR to remove steering converge checking from controllers.

What do you think about this opinion?

TakaHoribe commented:
@brkay54 Thank you for your PR.

It is reasonable not to use the stop steering command when the vehicle is trying to stop. However, I can't see the purpose of the move of the steering convergence check. What is the reason why you think the check should be in the gate?

brkay54 commented:
@TakaHoribe, thank you for the reply. 
> However, I can't see the purpose of the move of the steering convergence check. What is the reason why you think the check should be in the gate?
The main cause is https://github.com/autowarefoundation/autoware.universe/issues/4915
Checking steering convergence in the controller creates unnecessary complexity. For this case:

* pid_controller does not allow the vehicle to move because the steering is not converged.
* vehicle_cmd_gate does not allow to control steering because the vehicle stopped.
These two nodes wait for themselves forever.

Because vehicle_cmd_gate checks conditions for start and stopping, I think it makes sense to move steering checking into vehicle_cmd_gate.

TakaHoribe commented:
@brkay54 Thank you, I understood. It sounds reasonable. Would you reproduce the non-start issue in psim? Or does it happen only in the real vehicle testing?

brkay54 commented:
Yes, I can reproduce in psim.


https://github.com/autowarefoundation/autoware.universe/assets/45468306/d457b2d3-c3cc-40bc-9f47-4781abdca430
Sorry for the screen capture, there is a problem on my computer when recording screen. 

- Vehicle can start at the beginning in auto mode. 
- I set steering about 20 deg with manual controller.
- After I set 20 deg, when I clicked auto, vehicle can not move because of steering converging issue.
- I set steering about 0 def with manual controller.
- Vehicle can move in auto mode.

brkay54 commented:
@TakaHoribe -san, I rebased on the current master and solved the conflicts. And tested again, worked as we wanted.


https://github.com/autowarefoundation/autoware.universe/assets/45468306/19a2484b-9342-430d-a1e1-9957f49bdf7e

Merge order:
1 - https://github.com/autowarefoundation/autoware.universe/pull/5183 and https://github.com/autowarefoundation/autoware_launch/pull/607
2 - https://github.com/autowarefoundation/autoware.universe/pull/5189 and https://github.com/autowarefoundation/autoware_launch/pull/616
3 - https://github.com/autowarefoundation/autoware.universe/pull/5203

Thank you for your effort, if you have any concerns please let me know.

brkay54 commented:
@TakaHoribe -san, sorry to bother you, it is a friendly ping.

rej55 reviewed (approved):
LGTM

@TakaHoribe 
Could you approve this PR as a code owner?

rej55 commented on control/vehicle_cmd_gate/src/vehicle_cmd_gate.cpp:
```diff
@@ -430,13 +433,25 @@ void VehicleCmdGate::publishControlCommands(const Commands & commands)
+      if (!isSteeringConverged(filtered_commands.control.lateral)) {
+        filtered_commands.control.longitudinal = createStopTransitionControlCmd();
+      }
+    }
```
@brkay54 

```
 } else if (enable_keep_steering_until_convergence_) {
    if (fabs(current_kinematics_.twist.twist.linear.x) < stopped_vel_th) {
      if (!isSteeringConverged(filtered_commands.control.lateral)) {
        filtered_commands.control.longitudinal = createStopTransitionControlCmd();
      }
    }
 }
```

Do you have any reasons to implement this block into `vehicle_cmd_gate`?
I think it should be implemented at the end of `trajectory_follower_node` because `vehicle_cmd_gate` should just switch the control commands.

brkay54 replied:
Hi thank you for the review. Actually, I explained the reason above [here](https://github.com/autowarefoundation/autoware.universe/pull/5183#issuecomment-1794691616).

brkay54 replied:
Sorry to bother you, but could you check the comment [below](https://github.com/autowarefoundation/autoware.universe/pull/5183#issuecomment-1913727693)? @TakaHoribe @rej55

ahuazuipiaoliang commented:
@brkay54 @TakaHoribe @rej55 I checked the code, there are 3 relevant modules, which are trajectory_follower, vehicle_cmd_gate, default_ad_api;
1. If the required steering angle is too large, the steering angle cannot be converged, the the target velocity of longitudinal velocity should be set as 0; 
2. When the target velocity is 0, the vehicle_cmd_gate will output a state of `is_start_requested=false`;
3. If `is_start_requested==false`, the `default_ad_api` will change the state to `PAUSED` and keep calling the service `/control/vehicle_cmd_gate/set_pause` to set `PAUSED` state of `vehicle_cmd_gate`;
4. When the state of `vehicle_cmd_gate` is `PAUSED`, the `vehicle_cmd_gate` will keep sending `stop control cmd` which the steering angle is the `current_steering_angle` of the vehicle.

So the vehicle enters a dead loop and keeps `PAUSED` state.
I suggest create paused_control_cmd for `PAUSED` state in `vehicle_cmd_gate` that the `steering_cmd` is the command of `trajectory_follower`. When the steering angle is converged, the longitudinal controller can output normal control command.

brkay54 commented:
Firstly, we need to avoid sending `createStopControlCmd()`, while the vehicle is stopping (not stopped yet in disengage state). Because it causes sending current_steering while the vehicle is decelerating. You can see the problem [this issue](https://github.com/autowarefoundation/autoware.universe/issues/5796).

Thank you for the suggestion @ahuazuipiaoliang! It makes sense to me, however, it will cause an undesired behavior. If we change both `DISENGAGE` and `PAUSED` states to send control_cmd's steering command, the steering will be controllable all the time. So, the vehicle won't be fully stopped, always it listens to the control_cmd's steering command. cc @rej55 @TakaHoribe

rej55 commented:
@brkay54 cc @TakaHoribe 
Sorry for late reply, I agree for the suggestion from @ahuazuipiaoliang.
IMO, the trajectory_follower should send the control_cmd not to move the steering.
(@TakaHoribe said the same thing when I talked with him)

ahuazuipiaoliang commented:
@rej55 Thanks, Fujiyama. Actually, I have modified the code in my way and tested on real vehicle for several days. I have not found any issue yet.

brkay54 commented:
@rej55 Thanks for the reply, it makes sense to me. I am updating the PR.

brkay54 commented:
@rej55 @ahuazuipiaoliang Sorry to disturb you guys, thank you for your replies!

To be able to not listen to the trajectory_follower's output when the vehicle fully stopped, I found another way. I changed the structure like this:

```
  // Check engage
  if (!is_engaged_) {
    filtered_commands.control.longitudinal = createLongitudinalStopControlCmd();
  }

  // Check pause. Place this check after all other checks as it needs the final output.
  adapi_pause_->update(filtered_commands.control);
  if (adapi_pause_->is_paused()) {
    if (is_engaged_) {
      filtered_commands.control.longitudinal = createLongitudinalStopControlCmd();
    } else {
      filtered_commands.control = createStopControlCmd();
    }
  }
  ```
  
 In this way, the vehicle will be able to control steering if the vehicle is in PAUSED state. Also, it won't be able to control the steering when it is fully stopped and disengaged. I made some tests to check:
 

https://github.com/autowarefoundation/autoware.universe/assets/45468306/c0e0fb57-1a79-46a7-bdaa-7925b5ff2046

In the video above, I tested whether the vehicle can control the steering when we set a large steering value by using the manual controller or not. When I changed state to `AUTO`, the vehicle did not move until the steering converged as you can see.

In the second part, I set a large steering value again and I changed its state to `STOP`, as you can see, the vehicle did not change its steering. (It changed a little bit because the manual controller sets ENGAGE state to true while running and the steering was changing in the transition between ENGAGE to DISENGAGE)

Also, I checked the behavior of the vehicle while stopping. (It can control the steering or not while moving and disengaged) You can see the result in the video below:
 
https://github.com/autowarefoundation/autoware.universe/assets/45468306/0beac618-ceca-4ac2-8c2a-edec41320e86

As you can see, It still can control steering after I click the STOP button.

Also, I tried this structure too:

```
  // Check engage
  if (!is_engaged_) {
    filtered_commands.control.longitudinal = createLongitudinalStopControlCmd();
  }

  // Check pause. Place this check after all other checks as it needs the final output.
  adapi_pause_->update(filtered_commands.control);
  if (adapi_pause_->is_paused()) {
    filtered_commands.control.longitudinal = createLongitudinalStopControlCmd();
  }
  ```
  
It also solves the problems above, but, when the vehicle fully stopped, trajectory_follower can control the steering as you can see below:

https://github.com/autowarefoundation/autoware.universe/assets/45468306/c22c959c-fb50-4e8c-bc31-5894d6a5c1f6

Therefore I used the first structure. Please share your comments with me! Thank you for your efforts.

rej55 reviewed (approved):
LGTM.
Sorry for late review.

brkay54 commented:
@tkimura4 @TakaHoribe We need code-owner approve. Could you check the PR?

TakaHoribe reviewed (approved):
LGTM, thank you.
Note: the keep-stop command can send the target steering in order to control steering before moving.
